### PR TITLE
Pet.processTick

### LINKED ISSUE #278: Pet::processTick sucks
This method is such improperly written so I cannot even find the proper words to express it.

### PR DESCRIPTION
**Pull request by ForNeVeR**
Hereby I refactor the `Pet.processTick` method. It will no more use mutable variables and will use more easily extendable match system to manage pet states.

It will close #278.

### CODE REVIEW

**Comment by ForNeVeR** [comment]
cc @rexim, @folex

File: src/main/scala/ru/org/codingteam/horta/plugins/pet/Extractors.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package ru.org.codingteam.horta.plugins.pet
+
+object DyingPet {
+
+  def unapply(pet: PetData) = pet.health <= 0 || pet.satiation <= 0
```
Comment (rexim) [inline src/main/scala/ru/org/codingteam/horta/plugins/pet/Extractors.scala]: This is interesting. I always thought that `unapply` is supposed to return `Option[T]`.

File: src/main/scala/ru/org/codingteam/horta/plugins/pet/Pet.scala [hunk 2]
```diff
@@ -50,55 +40,58 @@ class Pet(roomId: String, location: ActorRef) extends Actor with ActorLogging {
     case GetPetDataInternal => sender ! petData
   }
 
-  private def processTick() = processAction { pet =>
+  private def nextState(pet: PetData, coinHolders: Iterable[String]) = {
```
Comment (rexim) [inline src/main/scala/ru/org/codingteam/horta/plugins/pet/Pet.scala]: Maybe some day we should consider to rewrite Pet using akka's become/unbecome?

File: src/main/scala/ru/org/codingteam/horta/plugins/pet/Pet.scala [hunk 3]
```diff
@@ -50,55 +40,58 @@ class Pet(roomId: String, location: ActorRef) extends Actor with ActorLogging {
...
+          sayToEveryone(random("%s's health is low.").format(nickname))
+        })
+
+      case _ => PetTickResponse(pet, noAction, { c => })
+    }
+  }
+
+  private def getPetVictimAsync(pet: PetData, coinHolders: Iterable[String]) = {
+    (location ? GetParticipants).mapTo[Protocol.ParticipantCollection].map { map =>
+      val possibleVictims = map.keys map ((x: String) => StringUtils.parseResource(x))
```
Comment (rexim) [inline src/main/scala/ru/org/codingteam/horta/plugins/pet/Pet.scala]: Is it possible to replace `(x: String) => StringUtils.parseResource(x)` with something like `StringUtils.parseResource(_)`?

File: src/main/scala/ru/org/codingteam/horta/plugins/pet/Extractors.scala (hunk 1 above)
Comment (ForNeVeR) [inline src/main/scala/ru/org/codingteam/horta/plugins/pet/Extractors.scala]: `Boolean` is possible too. Its purpose is obvious, see more at http://daily-scala.blogspot.ru/2009/10/boolean-extractors.html

File: src/main/scala/ru/org/codingteam/horta/plugins/pet/Pet.scala (hunk 2 above)
Comment (ForNeVeR) [inline src/main/scala/ru/org/codingteam/horta/plugins/pet/Pet.scala]: May be, my old friend, everything may be...

File: src/main/scala/ru/org/codingteam/horta/plugins/pet/Extractors.scala [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+}
+
+object HungerPet {
```
Comment (ForNeVeR) [inline src/main/scala/ru/org/codingteam/horta/plugins/pet/Extractors.scala]: That should be `HungryPet`.

File: src/main/scala/ru/org/codingteam/horta/plugins/pet/Extractors.scala (hunk 1 above)
Comment (rexim) [inline src/main/scala/ru/org/codingteam/horta/plugins/pet/Extractors.scala]: Yep, thanks for the link. :)

**Comment by ForNeVeR** [comment]
@rexim, I've made a little update. Please review again and say if we can publish that.

**Comment by rexim** [comment]
Looks good to me. I approve this.

**Comment by ForNeVeR** [comment]
mixer approved this shit mixer approved this shit

**Pull request merged**